### PR TITLE
Fix terminal restoration for shadowed tabs

### PR DESCRIPTION
**Pull request by hexa00**
The div used to open xterm.js needs to be visible when the xterm.js open
function is called so the size of the characters can be calculated using
getBoundingClientRect()

This patch makes it so that open is only called when the terminal is
visible.

This patch also saves/restores the terminal title so that the restoration
is smooth and makes sure that the resize callback is not called on a term
that was disposed.

Fixes: #818

Signed-off-by: Antoine Tremblay <antoine.tremblay@ericsson.com>

### CODE REVIEW

**Comment by hexa00** [comment]
FYI: I realised there was a possible bug if it waited for the backend terminal to be started but that the tab became unvisible meanwhile.

I just updated the PR to fix that.

File: packages/terminal/src/browser/terminal-widget.ts [hunk 1]
```diff
@@ -228,29 +236,79 @@ export class TerminalWidget extends BaseWidget implements StatefulWidget {
...
+        return Promise.resolve();
+    }
     protected createWebSocket(pid: string): WebSocket {
         const url = this.endpoint.getWebSocketUrl().resolve(pid);
         return this.webSocketConnectionProvider.createWebSocket(url.toString(), { reconnecting: false });
     }
 
-    protected onActivateRequest(msg: Message): void {
-        super.onActivateRequest(msg);
-        this.term.focus();
```
Comment (svenefftinge) [inline packages/terminal/src/browser/terminal-widget.ts]: Shouldn't the terminal still get focus on activation? I am thinking about a situation where I activate a terminal in another tab group which was shown before but just not active.

Comment (hexa00) [reply packages/terminal/src/browser/terminal-widget.ts]: Humm good point, I'll update it accordingly.

Comment (hexa00) [reply packages/terminal/src/browser/terminal-widget.ts]: Fixed in new version

**Comment by marcdumais-work** [comment]
@hexa00 Trying this patch I noticed a little glitch: If I start a few terminals and reload the browser, terminals that are not visible are not restored immediately (as designed), but also their title is not restored: it remains "<terminated>" until they become visible, at which time the correct title is applied.

**Comment by marcdumais-work** [comment]
peek 2017-11-30 06-56

**Comment by hexa00** [comment]
Seems like this happens on firefox but not on chrome, I'm checking it out

**Comment by hexa00** [comment]
Oops the title was changed to <terminated> on socket close.

Fixed now by removing that.

**Pull request merged**